### ISSUE TITLE
Binding Inheritance from outside passes to inside.

### ISSUE THREAD

**Issue by Kangz** [gpuweb/gpuweb#24]
Now that we agreed that compute and graphics work should be started and ended explicitly, an open question is whether resource bindings can be bound outside of "compute / render passes" and inherited inside these passes.

D3D12
-------

D3D12 doesn't have explicit bounds for graphics / compute work so be default root signature tables/descriptors/values are persistent. There are two root signatures, one for graphics and one for compute. Also the root signature is invalidated on a call to SetGraphicsRootSignature that changes the layout.

Overall when setting resource bindings outside of passes, we don't know if they should be put in the compute and graphics root signature, and they will in most cases be invalidated by a change to the root signature layout at the beginning of the pass. @kvark pointed out that in Vulkan (and all APIs) the functions binding resources know whether we bind to graphics or compute.

Metal
------

Bindings are set inside of the MTLComputeCommandEncoder and MTLRenderCommandEncoder and do not propagate outside of them.

Vulkan
------

Supports inheritance from outside render passes to inside.

Proposal
-------

Given that both D3D12 and Metal do not support inheritance, and given that vkCmdBindDescriptorSet is usually multiple order of magnitudes cheaper than vkCmdBeginRenderPass, it sounds ok to not have inheritance and ask the app to all the bindings inside the passes.

**Comment by kvark** [comment]
Thanks @Kangz for analysis!

## D3D12

> Overall when setting resource bindings outside of passes, we don't know if they should be put in the compute and graphics root signature

If we take Vulkan as the baseline, it's [vkCmdBindDescriptorSets](https://www.khronos.org/registry/vulkan/specs/1.0/man/html/vkCmdBindDescriptorSets.html) has a parameter saying whether these descriptors are set for compute or graphics.

> they will in most cases be invalidated by a change to the root signature layout at the beginning of the pass

The user may not necessarily change the root signature at the beginning of a pass. For example, they may re-use the same compute/graphics signatures across multiple passes in order to avoid resource re-binding.

## Vulkan

> Supports inheritance from outside render passes to inside.

I'd like to expand on that. Vulkan allows inheriting the resource bindings across [compatible](https://www.khronos.org/registry/vulkan/specs/1.0/html/vkspec.html#descriptorsets-compatibility) layouts. Thus, it make less sense to assume that re-binding all the resources at the beginning of compute/graphics pass is acceptable, since in Vulkan you'd rarely need to do that.

## Proposal

> Given that both D3D12 and Metal do not support inheritance

I wouldn't put D3D12 in the same bin as Metal here: D3D12 is fine with inheriting resource bindings across "render passes" as long as the root signature is unchanged. Thus, it does support inheritance, even if to lesser degree than Vulkan.

> given that vkCmdBindDescriptorSet is usually multiple order of magnitudes cheaper than vkCmdBeginRenderPass

This is true, although it's not clear if that sacrifice needs to happen. If we do inheritance of resource bindings, we'll see that the overhead is moved to `MTLRenderCommandEncoder` of Metal backend, where the same sort of justification may apply (e.g. binding resources is cheaper than starting a pass).

Another concern here is usability. Allowing an application to set descriptor sets outside of render passes would simply be more flexible, and thus allow richer set of abstractions to be build. For example, one may bind the layout/root signature followed by some global resources being set (e.g. weather parameters), only to proceed with a number of passes (e.g. shadow pass per light source) that only set their own specific resources.

TL;DR:
D3D12 model is a lot closer to Vulkan than it is to Metal, and believe providing the resource inheritance in the way of D3D12/Vulkan makes more sense.

**Comment by Kangz** [comment]
> If we take Vulkan as the baseline, it's vkCmdBindDescriptorSets has a parameter saying whether these descriptors are set for compute or graphics.

Thanks, I had completely forgotten about this.

> The user may not necessarily change the root signature at the beginning of a pass. For example, they may re-use the same compute/graphics signatures across multiple passes in order to avoid resource re-binding.

Theoretically an application may choose to do this, but I believe they won't because it trades a ton of flexibility to gain just a tiny bit of performance (as in something of the order of a usec of CPU + GPU time per frame).

> Another concern here is usability.

This is a good point, the resolution of this looks like it will be based on ergonomics: I think we agree that the performance impact of allowing or disallowing this inheritance is basically zero. The flip side if that no inheritance would mean the "BindDescriptorSet" function would know what "pipelineBindPoint" is, based on the wether we are in a compute or render pass. I don't have a strong opinion at this point and think we can defer this unless someone has a portability of performance concern.

**Comment by kvark** [comment]
Consensus has been reached to not have binding inheritance.
